pkg/agent/kubeagent.go

```diff
@@ -126,3 +169,265 @@ func (k *KubeAgent) GetFactories(_ context.Context) (otelcol.Factories, error) {
+// has changed.
+func (c *KubeAgentMonitor) ListenForConfigChanges(ctx context.Context) {
+
+	restartInterval, err := time.ParseDuration(c.ConfigCheckInterval)
+	if err != nil {
+		c.logger.Error(err.Error())
+	}
+
+	ticker := time.NewTicker(restartInterval)
+
+	for {
+		c.logger.Info("check for config changes after", zap.Duration("restartInterval", restartInterval))
+		select {
+		case <-ctx.Done():
+			// return
+		case <-ticker.C:
+			err = c.callRestartStatusAPI()
+			if err != nil {
+				c.logger.Info("error restarting agent on config change",
+					zap.Error(err))
+			}
+		}
+	}
+}
+
+// callRestartStatusAPI checks if there is an update in the otel-config at Middleware Backend
+// For a particular account
+func (c *KubeAgentMonitor) callRestartStatusAPI() error {
+
+	u, err := url.Parse(c.APIURLForConfigCheck)
+	if err != nil {
+		return err
+	}
+
+	baseURL := u.JoinPath(apiPathForRestart)
+	baseURL = baseURL.JoinPath(c.APIKey)
+	params := url.Values{}
+	params.Add("platform", "k8s")
+	params.Add("host_id", os.Getenv("MW_KUBE_CLUSTER_NAME"))
+	params.Add("cluster", os.Getenv("MW_KUBE_CLUSTER_NAME"))
+
+	// Add Query Parameters to the URL
+	baseURL.RawQuery = params.Encode() // Escape Query Parameters
+
+	resp, err := http.Get(baseURL.String())
+	if err != nil {
+		c.logger.Error("failed to call Restart-API", zap.String("url", baseURL.String()), zap.Error(err))
+		return err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		c.logger.Error("failed to call Restart-API", zap.Int("code", resp.StatusCode))
+		return ErrRestartStatusAPINotOK
+	}
+
+	var apiResponse apiResponseForRestart
+	if err := json.NewDecoder(resp.Body).Decode(&apiResponse); err != nil {
+		c.logger.Error("failed unmarshal Restart-API response", zap.Error(err))
+		return err
+	}
+
+	if apiResponse.Rollout.Daemonset {
+		c.logger.Info("restarting mw-agent")
+		if err := c.restartKubeAgent(DaemonSet); err != nil {
+			c.logger.Error("error restarting mw-agent daemonset", zap.Error(err))
+			return err
+		}
+	}
+
+	if apiResponse.Rollout.Deployment {
+		c.logger.Info("restarting mw-agent")
+		if err := c.restartKubeAgent(Deployment); err != nil {
+			c.logger.Error("error restarting mw-agent deployment", zap.Error(err))
+			return err
+		}
+	}
+
+	return err
+}
+
+// restartKubeAgent rewrites the configmaps and rollout restarts agent's data scraping components
+func (c *KubeAgentMonitor) restartKubeAgent(componentType ComponentType) error {
+
+	updateConfigMapErr := c.updateConfigMap(componentType)
+	if updateConfigMapErr != nil {
+		return updateConfigMapErr
+	}
+
+	rolloutRestartErr := c.rolloutRestart(componentType)
+	if rolloutRestartErr != nil {
+		return rolloutRestartErr
+	}
+
+	return nil
+}
+
+func (c *KubeAgentMonitor) SetClientSet() {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		c.logger.Error("Error getting in-cluster config", zap.String("error", err.Error()))
```

Comment:
Is this an error or just info ? If error, the code should be returning from here.